internal/cli/ops_manager_diagnose_archive_download.go

```diff
@@ -0,0 +1,99 @@
+// Copyright 2020 MongoDB Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cli
+
+import (
+	"io"
+	"os"
+
+	"go.mongodb.org/ops-manager/opsmngr"
+
+	"github.com/mongodb/mongocli/internal/description"
+	"github.com/mongodb/mongocli/internal/flags"
+	"github.com/mongodb/mongocli/internal/store"
+	"github.com/mongodb/mongocli/internal/usage"
+	"github.com/spf13/afero"
+	"github.com/spf13/cobra"
+)
+
+type opsManagerDiagnoseArchiveDownloadOpts struct {
+	globalOpts
+	out     string
+	limit   int64
+	minutes int64
+	fs      afero.Fs
+	store   store.ArchivesDownloader
+}
+
+func (opts *opsManagerDiagnoseArchiveDownloadOpts) initStore() error {
+	var err error
+	opts.store, err = store.New()
+	return err
+}
+
+func (opts *opsManagerDiagnoseArchiveDownloadOpts) Run() error {
+	out, err := opts.newWriteCloser()
+	if err != nil {
+		return err
+	}
+
+	if err := opts.store.DownloadArchive(opts.ProjectID(), opts.newDiagnosticsListOpts(), out); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (opts *opsManagerDiagnoseArchiveDownloadOpts) newWriteCloser() (io.WriteCloser, error) {
+	// Create file only if is not there already (don't overwrite)
+	ff := os.O_CREATE | os.O_TRUNC | os.O_WRONLY | os.O_EXCL
+	f, err := opts.fs.OpenFile(opts.out, ff, 0777)
+	return f, err
+}
+
+func (opts *opsManagerDiagnoseArchiveDownloadOpts) newDiagnosticsListOpts() *opsmngr.DiagnosticsListOpts {
+	return &opsmngr.DiagnosticsListOpts{
+		Limit:   opts.limit,
+		Minutes: opts.minutes,
+	}
+}
+
+// mongocli om diagnose-archive download [--out out] [--projectId projectId]
+func OpsManagerDiagnoseArchiveDownloadBuilder() *cobra.Command {
+	opts := &opsManagerDiagnoseArchiveDownloadOpts{
+		fs: afero.NewOsFs(),
+	}
+	cmd := &cobra.Command{
+		Use:     "download",
+		Aliases: []string{"get"},
+		Short:   description.DownloadDiagnoseArchive,
+		PreRunE: func(cmd *cobra.Command, args []string) error {
+			return opts.PreRunE(opts.initStore)
+		},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return opts.Run()
+		},
+	}
+
+	cmd.Flags().StringVarP(&opts.out, flags.Out, flags.OutShort, "", usage.LogOut)
+	cmd.Flags().Int64Var(&opts.limit, flags.Limit, 0, usage.ArchiveLimit)
+	cmd.Flags().Int64Var(&opts.minutes, flags.Minutes, 0, usage.ArchiveMinutes)
+
+	cmd.Flags().StringVar(&opts.projectID, flags.ProjectID, "", usage.ProjectID)
+
+	_ = cmd.MarkFlagRequired(flags.Out)
```

Comment:
Let's not make it mandatory but use diagnose-archive.tar.gz` as the default value of the flag

Reply:
If `out` were mandatory I would not know which name I should use to open the file in `newWriteCloser`
```
func (opts *opsManagerDiagnoseArchiveDownloadOpts) newWriteCloser() (io.WriteCloser, error) {
	// Create file only if is not there already (don't overwrite)
	ff := os.O_CREATE | os.O_TRUNC | os.O_WRONLY | os.O_EXCL
	f, err := opts.fs.OpenFile(opts.out, ff, 0777)
	return f, err
}
```

Reply:
Flags can have default values
`cmd.Flags().StringVarP(&opts.out, flags.Out, flags.OutShort, "diagnose-archive.tar.gz", usage.LogOut)`, sane defaults over a complicated API